experimental/sketching.py

```diff
@@ -0,0 +1,166 @@
+    out = np.sqrt(1 / self.rand_inds1.shape[-1]) * (x1fft * x2fft)
+    return np.concatenate((out.real, out.imag), 1)
+
+# Standard SRHT as a function
+def standardsrht(x, rand_inds, rand_signs):
+    xfft = np.fft.fftn(x * rand_signs, axes=(-1,))[:, rand_inds]
+    return np.sqrt(1 / rand_inds.shape[0]) * xfft
+
+@dataclasses.dataclass
+class PolyTensorSketch:
+    
+    rng: np.ndarray
+    
+    input_dim: int
+    sketch_dim: int
+    degree: int
+    
+    tree_rand_signs: Optional[list] = None
+    tree_rand_inds: Optional[list] = None
+    rand_signs: Optional[np.ndarray] = None
+    rand_inds: Optional[np.ndarray] = None
+    
+    replace = ...
+
+    def init_sketch(self) -> 'PolyTensorSketch':
+        
+        tree_rand_signs = [0 for i in range((self.degree - 1).bit_length())]
+        tree_rand_inds = [0 for i in range((self.degree - 1).bit_length())]
+        rng1, rng3 = random.split(self.rng, 2)
+    
+        ske_dim_ = self.sketch_dim // 4
+        deg_ = self.degree // 2
+        
+        for i in range((self.degree - 1).bit_length()):
+            rng1, rng2 = random.split(rng1)
+            
+            if i == 0:
+                tree_rand_signs[i] = random.choice(
+                    rng1, 2, shape=(deg_, 2, self.input_dim)) * 2 - 1
+                tree_rand_inds[i] = random.choice(rng2,
+                                                       self.input_dim,
+                                                       shape=(deg_, 2, ske_dim_))
+            else:
+                tree_rand_signs[i] = random.choice(
+                    rng1, 2, shape=(deg_, 2, ske_dim_)) * 2 - 1
+                tree_rand_inds[i] = random.choice(rng2,
+                                              ske_dim_,shape=(deg_, 2, ske_dim_))
+            deg_ = deg_ // 2
+        
+        rng1, rng2 = random.split(rng3,2)
+        rand_signs = random.choice(rng1, 2, shape=(self.degree * ske_dim_,)) * 2 - 1
+        rand_inds = random.choice(rng2,
+                                  self.degree * ske_dim_,shape=(self.sketch_dim // 2,))
+        
+        return self.replace(tree_rand_signs=tree_rand_signs,
+                            tree_rand_inds=tree_rand_inds,
+                            rand_signs=rand_signs,
+                            rand_inds=rand_inds)
+
+    
+    # TensorSRHT of degree 2
+    def tensorsrht(self, x1, x2, rand_inds, rand_signs):
+        x1fft = np.fft.fftn(x1 * rand_signs[0, :], axes=(-1,))[:, rand_inds[0, :]]
+        x2fft = np.fft.fftn(x2 * rand_signs[1, :], axes=(-1,))[:, rand_inds[1, :]]
+        return np.sqrt(1 / rand_inds.shape[1]) * (x1fft * x2fft)
+    
+    # Standard SRHT
+    def standardsrht(self, x, rand_inds, rand_signs):
+        xfft = np.fft.fftn(x * rand_signs, axes=(-1,))[:, rand_inds]
+        return np.sqrt(1 / rand_inds.shape[0]) * xfft
+
+    def sketch(self, x):
+        n = x.shape[0]
+        log_degree = len(self.tree_rand_signs)
+        V = [0 for i in range(log_degree)]
```

Comment:
Lets replace this with `np.zeros()` to have consistent type with following assignments, this should help with pytype checks.